rmw_fastrtps_shared_cpp/src/participant.cpp

```diff
@@ -161,19 +163,78 @@ rmw_fastrtps_shared_cpp::create_participant(
+        auto udp_transport = std::make_shared<eprosima::fastdds::rtps::UDPv4TransportDescriptor>();
+        udp_transport->maxInitialPeersRange = 119;
+        domainParticipantQos.transport().user_transports.push_back(udp_transport);
+        break;
+      }
+    case RMW_AUTOMATIC_DISCOVERY_RANGE_SUBNET:
+      // Nothing to do; use the default FastDDS behaviour
+      break;
+    case RMW_AUTOMATIC_DISCOVERY_RANGE_SYSTEM_DEFAULT:
+      // Nothing to do; use the default FastDDS behaviour
+      break;
+    default:
+      RMW_SET_ERROR_MSG("automatic_discovery_range is an unknown value");
+      return nullptr;
+      break;
+  }
 
-  if (localhost_only) {
-    // In order to use the interface white list, we need to disable the default transport config
-    domainParticipantQos.transport().use_builtin_transports = false;
-
-    // Add a UDPv4 transport with only localhost enabled
-    auto udp_transport = std::make_shared<eprosima::fastdds::rtps::UDPv4TransportDescriptor>();
-    udp_transport->interfaceWhiteList.emplace_back("127.0.0.1");
-    domainParticipantQos.transport().user_transports.push_back(udp_transport);
+  // Add initial peers if LOCALHOST or SUBNET are used
+  if (
+    RMW_AUTOMATIC_DISCOVERY_RANGE_LOCALHOST == discovery_options->automatic_discovery_range ||
+    RMW_AUTOMATIC_DISCOVERY_RANGE_SUBNET == discovery_options->automatic_discovery_range)
+  {
+    for (size_t ii = 0; ii < discovery_options->static_peers_count; ++ii) {
+      eprosima::fastrtps::rtps::Locator_t peer;
+      auto response = eprosima::fastrtps::rtps::IPLocator::resolveNameDNS(
+        discovery_options->static_peers[ii].peer_address);
+      // Get the first returned IPv4
+      if (response.first.size() > 0) {
+        eprosima::fastrtps::rtps::IPLocator::setIPv4(peer, response.first.begin()->data());
+      } else {
+        RMW_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Unable to resolve peer %s\n",
+          discovery_options->static_peers[ii].peer_address);
+        return nullptr;
+      }
+      // Not specifying the port of the peer means FastDDS will try all
+      // possible participant ports according to the port calculation equation
+      // in the RTPS spec section 9.6.1.1, up to the number of peers specified
+      // in maxInitialPeersRange.
+      domainParticipantQos.wire_protocol().builtin.initialPeersList.push_back(peer);
+    }
+  }
 
-    // Add SHM transport if available
-    auto shm_transport = std::make_shared<eprosima::fastdds::rtps::SharedMemTransportDescriptor>();
-    domainParticipantQos.transport().user_transports.push_back(shm_transport);
+  if (RMW_AUTOMATIC_DISCOVERY_RANGE_LOCALHOST == discovery_options->automatic_discovery_range) {
+    // Add localhost as a static peer
+    eprosima::fastrtps::rtps::Locator_t peer;
+    eprosima::fastrtps::rtps::IPLocator::setIPv4(peer, "127.0.0.1");
+    domainParticipantQos.wire_protocol().builtin.initialPeersList.push_back(peer);
```

Comment:
For the case of `RMW_AUTOMATIC_DISCOVERY_RANGE_SUBNET`, if `initialPeersList` is not empty, we should add the default multicast address to it.

Reply:
To the initial peer list, or as a multicast locator? I assumed the latter in cf9bf83